pinot-broker/src/main/java/org/apache/pinot/broker/routing/HelixExternalViewBasedTimeBoundaryService.java

```diff
@@ -48,77 +50,106 @@ public HelixExternalViewBasedTimeBoundaryService(ZkHelixPropertyStore<ZNRecord>
   }
 
   public void updateTimeBoundaryService(ExternalView externalView) {
-    if (_propertyStore == null) {
-      return;
-    }
-    String tableName = externalView.getResourceName();
-    // Do nothing for realtime table.
-    if (TableNameBuilder.getTableTypeFromTableName(tableName) == TableType.REALTIME) {
+    String tableNameWithType = externalView.getResourceName();
+
+    // Skip real-time table, only use offline table to update the time boundary
+    if (TableNameBuilder.getTableTypeFromTableName(tableNameWithType) == TableType.REALTIME) {
       return;
     }
 
     Set<String> offlineSegmentsServing = externalView.getPartitionSet();
     if (offlineSegmentsServing.isEmpty()) {
-      LOGGER.info("Skipping updating time boundary service for table '{}' with no offline segments.", tableName);
+      LOGGER.warn("Skipping updating time boundary for table: '{}' with no offline segment", tableNameWithType);
       return;
     }
 
-    TableConfig offlineTableConfig = ZKMetadataProvider.getOfflineTableConfig(_propertyStore, tableName);
-    assert offlineTableConfig != null;
-    TimeUnit tableTimeUnit = offlineTableConfig.getValidationConfig().getTimeType();
-    if (tableTimeUnit == null) {
-      LOGGER.info("Skipping updating time boundary service for table '{}' because time unit is not set", tableName);
+    // TODO: when we start using dateTime, pick the time column from the retention config, and use the DateTimeFieldSpec
+    //       from the schema to determine the time unit
+    // TODO: support SDF
+    TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
+    assert tableConfig != null;
+    SegmentsValidationAndRetentionConfig retentionConfig = tableConfig.getValidationConfig();
+    String timeColumn = retentionConfig.getTimeColumnName();
+    TimeUnit tableTimeUnit = retentionConfig.getTimeType();
+    if (timeColumn == null || tableTimeUnit == null) {
+      LOGGER.error("Skipping updating time boundary for table: '{}' because time column/unit is not set",
+          tableNameWithType);
       return;
     }
 
-    // Bulk reading all segment zk-metadata at once is more efficient than reading one at a time.
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, tableNameWithType);
```

Comment:
This is for the TODO in line 66. Because segment is generated based on the schema, we should always pick the time column and unit from schema instead of the table config. In the future if we have multiple time columns (DateTimeField) in schema, we can use the time column from table config to determine which one to use as the main time column, but the time unit should still from schema.
Because we want to follow the time field spec in schema in the future, I'll change it to use time column/unit from schema, and keep the checks so that we can find the inconsistency.